explore/explore.h

```diff
@@ -0,0 +1,152 @@
+   * @param pdf_first Iterator pointing to the pre-allocated beginning of the pdf to be generated by this function.
+   * @param pdf_last Iterator pointing to the pre-allocated end of the pdf to be generated by this function.
+   * @return int returns 0 on success, otherwise an error code as defined by E_EXPLORATION_*. 
+   */
+  template<typename InputIt, typename OutputIt>
+  int generate_softmax(float lambda, InputIt scores_begin, InputIt scores_last, OutputIt pdf_first, OutputIt pdf_last);
+
+  /**
+   * @brief Generates an exploration distribution according to votes on actions.
+   * 
+   * @tparam InputIt Iterator type of the input actions. Must be an InputIterator.
+   * @tparam OutputIt Iterator type of the pre-allocated pdf. Must be a RandomAccessIterator.
+   * @param top_actions_begin Iterator pointing to the beginning of the top actions.
+   * @param top_actions_last Iterator pointing to the end of the top actions.
+   * @param pdf_first Iterator pointing to the pre-allocated beginning of the pdf to be generated by this function.
+   * @param pdf_last Iterator pointing to the pre-allocated end of the pdf to be generated by this function.
+   * @return int returns 0 on success, otherwise an error code as defined by E_EXPLORATION_*. 
+   */
+  template<typename InputIt, typename OutputIt>
+  int generate_bag(InputIt top_actions_begin, InputIt top_actions_last, OutputIt pdf_first, OutputIt pdf_last);
+
+  /**
+   * @brief Updates the pdf to ensure each action is explored with at least minimum_uniform/num_actions.
+   * 
+   * @tparam It Iterator type of the pdf. Must be a RandomAccessIterator.
+   * @param minimum_uniform The minimum amount of uniform distribution to impose on the pdf.
+   * @param update_zero_elements If true elements with zero probability are updated, otherwise those actions will be unchanged.
+   * @param pdf_first Iterator pointing to the pre-allocated beginning of the pdf to be generated by this function.
+   * @param pdf_last Iterator pointing to the pre-allocated end of the pdf to be generated by this function.
+   * @return int returns 0 on success, otherwise an error code as defined by E_EXPLORATION_*. 
+   */
+  template<typename It>
+  int enforce_minimum_probability(float minimum_uniform, bool update_zero_elements, It pdf_first, It pdf_last);
+
+  /**
+   * @brief Sample an index from the provided pdf.
+   * 
+   * @tparam InputIt Iterator type of the pdf. Must be an InputIterator.
+   * @param seed The seed for the pseudo-random generator.
+   * @param pdf_first Iterator pointing to the beginning of the pdf.
+   * @param pdf_last Iterator pointing to the end of the pdf.
+   * @param chosen_index returns the chosen index.
+   * @param pdf_updated set to true if the pdf required normalization and was updated in-place.
+   * @return int returns 0 on success, otherwise an error code as defined by E_EXPLORATION_*. 
+   */
+  template<typename InputIt>
```

Comment:
I believe this should be an It not a InputIt.

Reply:
done